include/mbgl/style/layer.hpp

```diff
@@ -65,6 +66,14 @@ struct LayerTypeInfo {
     const enum class TileKind : uint8_t { Geometry, Raster, RasterDEM, NotRequired } tileKind;
 };
 
+struct LayerProperty {
+    enum class Kind : uint8_t { Undefined, Constant, Expression, Transition };
+    LayerProperty(Value value_, Kind kind_) : value(std::move(value_)), kind(kind_) {}
+    LayerProperty() = default;
+    const Value value;
+    const Kind kind = Kind::Undefined;
```

Comment:
Instead of having templated `makeLayerProperty`, we could instead make `LayerProperty` constructors private and provide factory functions for each kind instead.

Reply:
I'd love to keep `make..` functions inside converter and keep them free functions, so that it's easier to add new ones.